File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-3.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        count = 0
        res = -1
        
        def inorder(node):
            nonlocal count
            nonlocal res

            if not node:
                return False

            if inorder(node.left):
                return True

            count += 1
            if count == k:
                res = node.val
            
            if inorder(node.right):
                return True

            return False

        inorder(root)

        return res

File: Data Structures & Algorithms/binary-tree-postorder-traversal/submission-3.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    /**
     * @param {TreeNode} root
     * @return {number[]}
     */
    postorderTraversal(root: TreeNode | null): number[] {
        const res = []
        this.postorder(root, res)
        return res
    }

    postorder(node: TreeNode | null, res: number[]): void {
        if (!node) return

        this.postorder(node.left, res)
        this.postorder(node.right, res)
        res.push(node.val)
    }
}
